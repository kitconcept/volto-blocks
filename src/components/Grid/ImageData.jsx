import React from 'react';
import PropTypes from 'prop-types';
import { Accordion, Form, Grid, Segment } from 'semantic-ui-react';
import {
  defineMessages,
  FormattedMessage,
  injectIntl,
  intlShape,
} from 'react-intl';
import { CheckboxWidget, Icon, TextWidget } from '@plone/volto/components';
import { settings } from '~/config';
import { AlignTile, flattenToAppURL } from '@plone/volto/helpers';
import { compose } from 'redux';
import withObjectBrowser from '@plone/volto/components/manage/Sidebar/ObjectBrowser';

import imageSVG from '@plone/volto/icons/image.svg';
import clearSVG from '@plone/volto/icons/clear.svg';
import upSVG from '@plone/volto/icons/up-key.svg';
import downSVG from '@plone/volto/icons/down-key.svg';
import navTreeSVG from '@plone/volto/icons/nav.svg';

const messages = defineMessages({
  Source: {
    id: 'Source',
    defaultMessage: 'Source',
  },
  Image: {
    id: 'Image',
    defaultMessage: 'Image',
  },
  Origin: {
    id: 'Origin',
    defaultMessage: 'Origin',
  },
  AltText: {
    id: 'Alt text',
    defaultMessage: 'Alt text',
  },
  Align: {
    id: 'Alignment',
    defaultMessage: 'Alignment',
  },
  LinkTo: {
    id: 'Link to',
    defaultMessage: 'Link to',
  },
  openLinkInNewTab: {
    id: 'Open in a new tab',
    defaultMessage: 'Open in a new tab',
  },
  NoImageSelected: {
    id: 'No image selected',
    defaultMessage: 'No image selected',
  },
  externalURL: {
    id: 'External URL',
    defaultMessage: 'External URL',
  },
});

const ImageData = ({
  data,
  tile,
  onChangeTile,
  openObjectBrowser,
  required = false,
  intl,
}) => {
  const [alt, setAlt] = React.useState(data.alt || '');
  const [activeAccIndex, setActiveAccIndex] = React.useState(0);

  function handleAccClick(e, titleProps) {
    const { index } = titleProps;
    const newIndex = activeAccIndex === index ? -1 : index;

    setActiveAccIndex(newIndex);
  }

  return (
    <>
      {data.url && (
        <Segment className="sidebar-metadata-container" secondary>
          {data.url.split('/').slice(-1)[0]}
          {data.url.includes(settings.apiPath) && (
            <img
              src={`${flattenToAppURL(data.url)}/@@images/image/mini`}
              alt={alt}
            />
          )}
          {!data.url.includes(settings.apiPath) && (
            <img src={data.url} alt={alt} style={{ width: '50%' }} />
          )}
        </Segment>
      )}
      <Segment className="form sidebar-image-data">
        <TextWidget
          id="source"
          title={intl.formatMessage(messages.Source)}
          required={false}
          value={data.url ? flattenToAppURL(data.url) : ''}
          icon={data.url ? clearSVG : navTreeSVG}
          iconAction={
            data.url
              ? () => {
                  onChangeTile(tile, {
                    ...data,
                    url: '',
                  });
                }
              : () => openObjectBrowser()
          }
          onChange={(name, value) => {
            onChangeTile(tile, {
              ...data,
              url: value,
            });
          }}
        />

        <TextWidget
          id="alt"
          title={intl.formatMessage(messages.AltText)}
          required={false}
          value={alt}
          onChange={(name, value) => {
            onChangeTile(tile, {
              ...data,
              alt: value,
            });
            setAlt(value);
          }}
        />

        <TextWidget
          id="link"
          title={intl.formatMessage(messages.LinkTo)}
          required={false}
          value={data.href}
          icon={data.href ? clearSVG : navTreeSVG}
          iconAction={
            data.href
              ? () => {
                  onChangeTile(tile, {
                    ...data,
                    href: '',
                  });
                }
              : () => openObjectBrowser('link')
          }
          onChange={(name, value) => {
            onChangeTile(tile, {
              ...data,
              href: value,
            });
          }}
        />
        <CheckboxWidget
          id="openLinkInNewTab"
          title={intl.formatMessage(messages.openLinkInNewTab)}
          value={data.openLinkInNewTab ? data.openLinkInNewTab : false}
          onChange={(name, value) => {
            onChangeTile(tile, {
              ...data,
              openLinkInNewTab: value,
            });
          }}
        />
      </Segment>
    </>
  );
};

ImageData.propTypes = {
  data: PropTypes.objectOf(PropTypes.any).isRequired,
  tile: PropTypes.string.isRequired,
  onChangeTile: PropTypes.func.isRequired,
  openObjectBrowser: PropTypes.func.isRequired,
  intl: intlShape.isRequired,
};

export default compose(
  withObjectBrowser,
  injectIntl,
)(ImageData);
