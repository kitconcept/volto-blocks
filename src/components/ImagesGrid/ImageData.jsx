import React from 'react';
import PropTypes from 'prop-types';
import { useIntl } from 'react-intl';
import config from '@plone/volto/registry';
import withObjectBrowser from '@plone/volto/components/manage/Sidebar/ObjectBrowser';
import { SchemaRenderer } from '../../components';
import { ImagesGridSchema } from './schema';

const ImageData = (props) => {
  const { data, dataGrid, block, onChangeBlock } = props;

  const intl = useIntl();

  const schema = ImagesGridSchema({ ...props, intl });

  const applyVariationSchemaExtender = (schema) => {
    const variations =
      config.blocks?.blocksConfig?.[dataGrid['@type']]?.variations;

    const schemaExtender =
      variations?.[dataGrid?.variation]?.['schemaExtenderItem'];

    if (schemaExtender) {
      return schemaExtender(schema, props, intl);
    } else {
      return schema;
    }
  };

  return (
    <SchemaRenderer
      schema={applyVariationSchemaExtender(schema)}
      title={schema.title}
      onChangeField={(id, value) => {
        onChangeBlock(block, {
          ...data,
          [id]: value,
        });
      }}
      formData={data}
      fieldIndex={data.index}
    />
  );
};

ImageData.propTypes = {
  data: PropTypes.objectOf(PropTypes.any).isRequired,
  block: PropTypes.string.isRequired,
  onChangeBlock: PropTypes.func.isRequired,
  openObjectBrowser: PropTypes.func.isRequired,
};

export default withObjectBrowser(ImageData);
