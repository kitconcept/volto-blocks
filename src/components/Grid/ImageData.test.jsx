import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-intl-redux';

import ImageData from './ImageData';

const mockStore = configureStore();

test('renders an Image Data Sidebar component', () => {
  const store = mockStore({
    intl: {
      locale: 'en',
      messages: {},
    },
  });
  const component = renderer.create(
    <Provider store={store}>
      <ImageData
        id="dcdf1f42-645d-48f6-9531-357bdc2e1881"
        data={{
          '@type': 'imagesGrid',
          columns: [
            {
              '@type': 'image',
              href: '/news/a-news-item-that-tells-things',
              id: '2110a241-1389-4cda-8811-77031a540efa',
              index: 0,
            },
            {
              '@type': 'image',
              href: '/news/another-news-item',
              id: '3ccd6338-4d13-4f22-a68c-14b6b60cf860',
              index: 1,
            },
            {
              '@type': 'image',
              id: '2ac3b8d1-3ea8-4d78-a4ce-b2997a961c5e',
            },
            {
              '@type': 'image',
              id: 'e947b1e5-b9b1-4fc1-acf3-cb04f78a3574',
            },
          ],
        }}
        type="imagesGrid"
        tile="1234"
        pathname="/news"
        onChangeTile={() => {}}
        openObjectBrowser={() => {}}
      />
    </Provider>,
  );
  const json = component.toJSON();
  expect(json).toMatchSnapshot();
});
