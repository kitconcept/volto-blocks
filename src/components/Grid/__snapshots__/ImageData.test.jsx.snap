// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an Image Data Sidebar component 1`] = `""`;
